bigchaindb/tendermint/lib.py

```diff
@@ -347,12 +351,18 @@ def validate_transaction(self, tx, current_transactions=[]):
             except ValidationError as e:
                 logger.warning('Invalid transaction (%s): %s', type(e).__name__, e)
                 return False
+        return transaction.validate(self, current_transactions)
+
+    def is_valid_transaction(self, tx, current_transactions=[]):
+        # The below exception handling is quite broad because this method
+        # is called by ABCI proxy app. An exception not caught in this method
+        # will cause a crash in the ABCI server leaving the system in a
+        # dysfunctional state
         try:
-            return transaction.validate(self, current_transactions)
-        except ValidationError as e:
```

Comment:
`ValidationError` doesn't handle `CriticalDoubleSpend` and this causes a crash. Both these exception types have a common parent `BigchainBDError`.

Reply:
I think we should investigate why there was a `CriticalDoubleSpend`.

The exception is defines as:
```python
class CriticalDoubleSpend(BigchainDBError):
    """Data integrity error that requires attention"""
```

It is raised in [`get_spent`](https://github.com/bigchaindb/bigchaindb/blob/9cdd46c0844b442af738e83c89dcc24fae8d21c7/bigchaindb/core.py#L389). Do you have an idea on why this is happening?

Reply:
Perhaps, there is a test reproducing it? Should `validate_transaction` never raise `BigchainBDError`?

Reply:
@vrde My understanding is that [`get_spent`](https://github.com/bigchaindb/bigchaindb/blob/master/bigchaindb/tendermint/lib.py#L241) is called with `current_transactions` from the un-commited block which probably contains duplicate transactions which raises an exception.

Reply:
How can we have an un-commited block with duplicate transactions?

Reply:
@ldmberman this [test](https://github.com/bigchaindb/bigchaindb/blob/master/tests/tendermint/test_core.py#L105) check that.

@codegeschrei During the Tendermint [state machine](http://tendermint.readthedocs.io/projects/tools/en/master/introduction.html#intro-to-abci) until `commit` is executed the block is considered un-commited. If a transaction T2 spends T1 and there are two instances of T2 transaction included in the same block then we have a duplicate transactions in the same block.

Reply:
But we should reject a double spend right in the beginning, or am I wrong?

Reply:
@codegeschrei You are right. The initial check for double spend compares the incoming transactions to the transactions which have been commited to the chain. It doesn't compare to the transaction that haven't yet been commited and are still in the mempool (this check is perfomed during the `deliver_tx` ABCI call)